### PR TITLE
Potential perf improvements

### PR DESCRIPTION
**Pull request by islijepcevic**
Update `encode_sequence` to return indices instead of dense one-hot matrix.

* Dense one-hot matrix was created in a python for-loop. We could use methods from pytorch (or sklearn) to generate the matrix more optimally.
* Instead of dense matrix, we could use a list of indices. Then, instead of linear layers, we can use Embedding layer. We do that here.

In a direct comparison, there is a statistically significant improvement (in ms). When running entire training, there is not much clear difference in iterations per second.

### CODE REVIEW

File: rnafold/utils/spatial/construct_structure.py [hunk 1]
```diff
@@ -168,7 +169,8 @@ def _attach_sidechain(self, global_atom_coordinates, seq_encoded, n_transform):
 
         sc_global_coords = n_transform.apply(self.sc_local_coords.view(-1, 1, 1, 3)) # RA x B x S x 3
         sc_global_coords = sc_global_coords.permute(1, 2, 0, 3) #  B x S x RA x 3
-        sc_global_coords = seq_encoded[..., None, None] * sc_global_coords.view(bs, seq_len, NUM_RESIDUE_TOKENS, NUM_ATOM_TYPES, 3) # B x S x R x A x 3
+        seq_onehot = F.one_hot(seq_encoded, num_classes=NUM_RESIDUE_TOKENS)
+        sc_global_coords = seq_onehot[..., None, None] * sc_global_coords.view(bs, seq_len, NUM_RESIDUE_TOKENS, NUM_ATOM_TYPES, 3) # B x S x R x A x 3
```
Comment (islijepcevic) [inline rnafold/utils/spatial/construct_structure.py:173]: Ideally, we'd use something like `torch.index_select()` instead of dense onehot, but couldn't find a batched version of `index_select`. So, onehot is probably better than a python for-loop over the batch indices.

I also tried using sparse tensors here, wasn't clearly better, so opted to keep simplicity.
Sparse code like this:
```
        sc_collapsed = sc_global_coords.reshape((bs, seq_len, NUM_RESIDUE_TOKENS, -1))

        indices_3d = torch.stack(
            # Batch index
            torch.repeat_interleave(torch.arange(bs, device=seq_encoded.device), seq_len),
            # Seq index
            torch.arange(seq_len, device=seq_encoded.device).repeat(bs),
            # Atom index
            seq_encoded.view(-1),
        )
        seq_onehot = torch.sparse_coo_tensor(
            indices=indices_3d,
            values=torch.ones(bs * seq_len, device=seq_encoded.device).long(),
            size=torch.Size((bs, seq_len, NUM_RESIDUE_TOKENS)),
            device=seq_encoded.device,
        )
        # Convert the COO sparse tensor to CSR format for efficient operations
        seq_onehot = seq_onehot.coalesce().to_sparse()

        sc_global_coords = seq_onehot[..., None] * sc_collapsed
        sc_global_coords = sc_global_coords.view(bs, seq_len, NUM_RESIDUE_TOKENS, NUM_ATOM_TYPES, 3)  # B x S x R x A x 3
        sc_global_coords = sc_global_coords.sum(dim=-3) # B x S x A x 3
```

Comment (RJPenic) [reply rnafold/utils/spatial/construct_structure.py:173]: Let's keep it like it is for now, I would like to keep this part of the code as simple as possible. Also, I would try to avoid using sparse tensors since they are still in beta so it is very possible that their functionalities and API change in the near future.

File: rnafold/utils/seq_utils.py [hunk 2]
```diff
@@ -1,20 +1,20 @@
...
-        seq_encoded[i, RESIDUE_TOKEN_ORDER[tkn]] = 1.
-
-    return seq_encoded
+def encode_sequence(seq: str) -> torch.Tensor:
+    return torch.tensor([
+        RESIDUE_TOKEN_ORDER[res]
+        if res in STANDARD_RESIDUE_TOKENS
+        else RESIDUE_TOKEN_ORDER[UNDEFINED_TKN]
+        for res in seq
+    ])
```
Comment (RJPenic) [inline rnafold/utils/seq_utils.py:20]: Looks good!

**Review by RJPenic — APPROVED** [review]
Looks good! I think these changes won't cause any major speed-ups during training since sequence encoding part is executed in dataloader workers but the code looks cleaner. Feel free to merge if you don't have anything else to add. 😄

File: rnafold/utils/spatial/construct_structure.py (hunk 1 above)
Comment (islijepcevic) [reply rnafold/utils/spatial/construct_structure.py:173]: Just to clarify, should I keep the original code, or code that is in this PR (in current state)?

File: rnafold/layers/feature_preprocessing.py [hunk 3]
```diff
@@ -17,20 +17,20 @@ def __init__(self, c_m, c_z, sec_struct_bins=20, splash_bins=5, shape_bins=10):
         self.shape_binner = Binner(num_bins=shape_bins + 1, bin_size=0.15, first_bin_upper_bound=-0.1)
         self.splash_binner = Binner(num_bins=splash_bins, bin_size=1. / splash_bins)
 
-        self.linear_a, self.linear_b = nn.Linear(NUM_RESIDUE_TOKENS, c_z), nn.Linear(NUM_RESIDUE_TOKENS, c_z)
+        self.embedding_a, self.embedding_b = nn.Embedding(NUM_RESIDUE_TOKENS, c_z), nn.Embedding(NUM_RESIDUE_TOKENS, c_z)
```
Comment (RJPenic) [inline rnafold/layers/feature_preprocessing.py:20]: This won't work properly. I just remembered that we don't have padding token defined, to use embeddings you would need to modify batch preparation in data module.

File: rnafold/data/constants.py [hunk 4]
```diff
@@ -27,11 +27,13 @@
 GUANINE_TKN  = 'G'
 CYTOSINE_TKN = 'C'
 UNDEFINED_TKN = 'N'
+PADDING_TKN = '-'
```
Comment (RJPenic) [inline rnafold/data/constants.py:30]: Maybe put `<pad>` (or something similar) instead because `-` can represent [gaps in FASTA files](https://en.wikipedia.org/wiki/FASTA_format). Just to avoid potential confusions. 😄

File: rnafold/layers/feature_preprocessing.py [hunk 5]
```diff
@@ -17,12 +17,12 @@ def __init__(self, c_m, c_z, sec_struct_bins=20, splash_bins=5, shape_bins=10):
         self.shape_binner = Binner(num_bins=shape_bins + 1, bin_size=0.15, first_bin_upper_bound=-0.1)
         self.splash_binner = Binner(num_bins=splash_bins, bin_size=1. / splash_bins)
 
-        self.embedding_a, self.embedding_b = nn.Embedding(NUM_RESIDUE_TOKENS, c_z), nn.Embedding(NUM_RESIDUE_TOKENS, c_z)
+        self.embedding_a, self.embedding_b = nn.Embedding(NUM_RESIDUE_VOCAB, c_z, padding_idx=0), nn.Embedding(NUM_RESIDUE_VOCAB, c_z, padding_idx=0)
```
Comment (RJPenic) [inline rnafold/layers/feature_preprocessing.py:20]: Just in case padding index isn't zero:
```suggestion
        self.embedding_a, self.embedding_b = nn.Embedding(NUM_RESIDUE_VOCAB, c_z, padding_idx=RESIDUE_VOCAB_ORDER[PADDING_TOKEN]), nn.Embedding(NUM_RESIDUE_VOCAB, c_z, padding_idx=RESIDUE_VOCAB_ORDER[PADDING_TOKEN])
```

File: rnafold/layers/feature_preprocessing.py [hunk 6]
```diff
@@ -17,12 +17,12 @@ def __init__(self, c_m, c_z, sec_struct_bins=20, splash_bins=5, shape_bins=10):
...
         self.linear_sec_struct = nn.Linear(sec_struct_bins, c_z)
         self.linear_shape = nn.Linear(shape_bins + 1, c_m)
         self.linear_splash = nn.Linear(splash_bins, c_z)
         
-        self.embedding_seq = nn.Embedding(NUM_RESIDUE_TOKENS, c_m)
+        self.embedding_seq = nn.Embedding(NUM_RESIDUE_VOCAB, c_m, padding_idx=0)
```
Comment (RJPenic) [inline rnafold/layers/feature_preprocessing.py:25]: Same thing regarding the padding index here:
```suggestion
        self.embedding_seq = nn.Embedding(NUM_RESIDUE_VOCAB, c_m, padding_idx=RESIDUE_VOCAB_ORDER[PADDING_TOKEN])
```

File: rnafold/utils/spatial/construct_structure.py [hunk 7]
```diff
@@ -46,56 +46,56 @@
 
 # Local side coordinates
 # Sidechain
-SC_LOCAL_COORDS = np.full((NUM_RESIDUE_TOKENS, NUM_ATOM_TYPES, 3), fill_value=0.0, dtype=np.float32)
+SC_LOCAL_COORDS = np.full((NUM_RESIDUE_VOCAB, NUM_ATOM_TYPES, 3), fill_value=0.0, dtype=np.float32)
```
Comment (RJPenic) [inline rnafold/utils/spatial/construct_structure.py:49]: I think this should work fine even with the expanded vocabulary (nucleotides + unknown token + padding token).

File: rnafold/data/constants.py (hunk 4 above)
Comment (islijepcevic) [reply rnafold/data/constants.py:30]: makes sense. I don't think we ever convert tokens back to strings, do we?
In case we do, maybe our parsing won't work because parsing assumes every token is individual char?

File: rnafold/utils/spatial/construct_structure.py (hunk 7 above)
Comment (islijepcevic) [reply rnafold/utils/spatial/construct_structure.py:49]: thanks  for checking
